nodes/arm_node.cpp

```diff
@@ -1,110 +1,247 @@
+#include <ros/console.h>
+#include <ros/package.h>
+
+#include <actionlib/server/simple_action_server.h>
+
 #include <geometry_msgs/Point.h>
+
 #include <hebi_cpp_api_examples/TargetWaypoints.h>
+#include <hebi_cpp_api_examples/ArmMotionAction.h>
+
+#include "group_command.hpp"
 
 #include "robot/arm.hpp"
 #include "robot_model.hpp"
 
-#include <ros/console.h>
-#include <ros/package.h>
-
 namespace hebi {
-  namespace ros {
-
-    class ArmNode {
-    public:
-      ArmNode(arm::Arm& arm)
-        : arm_(arm)
-      { }
-
-      // "Jog" the target end effector location in (x,y,z) space, replanning
-      // smoothly to the new location
-      void offsetTargetCallback(geometry_msgs::Point data) {
-        
-        // Only update if target changes!
-        if (data.x == 0 && data.y == 0 && data.z == 0)
-          return;
-
-        // Initialize target from feedback as necessary
-        if (!isTargetInitialized()) {
-          auto pos = arm_.getKinematics().FK(arm_.getLastFeedback().getPosition());
-          target_xyz_.x() = pos.x();
-          target_xyz_.y() = pos.y();
-          target_xyz_.z() = pos.z();
-        }
-
-        // Update the target point
-        target_xyz_.x() += data.x / 100.0; // Converts to cm
-        target_xyz_.y() += data.y / 100.0; // Converts to cm
-        target_xyz_.z() += data.z / 100.0; // Converts to cm
-         
-        Eigen::VectorXd ik_result_joint_angles =
-          arm_.getKinematics().solveIK(arm_.getLastFeedback().getPosition(),
-          target_xyz_);
-
-        // Replan:
-        arm_.getTrajectory().replan(
-          ::ros::Time::now().toSec(),
-          arm_.getLastFeedback(),
-          ik_result_joint_angles);
+namespace ros {
+
+class ArmNode {
+public:
+  ArmNode(arm::Arm& arm) : arm_(arm) { }
+
+  // Replan a smooth joint trajectory from the current location through a
+  // series of cartesian waypoints.
+  // xyz positions should be a 3xn vector of target positions
+  void updateCartesianWaypoints(const Eigen::Matrix3Xd& xyz_positions,
+      const Eigen::Matrix3Xd& end_tip_directions) {
+    // Data sanity check:
+    if (end_tip_directions.cols() > 0 && end_tip_directions.cols() != xyz_positions.cols())
+      return;
+
+    // Update stored target position:
+    target_xyz_ = xyz_positions.col(xyz_positions.cols() - 1);
+
+    // These are the joint angles that will be added
+    auto num_waypoints = xyz_positions.cols();
+    Eigen::MatrixXd positions(arm_.size(), num_waypoints);
+
+    // Plan to each subsequent point from the last position
+    // (We use the last position command for smoother motion)
+    Eigen::VectorXd last_position = arm_.getLastFeedback().getPositionCommand();
+
+    // Get joint angles to move to each waypoint
+    for (size_t i = 0; i < num_waypoints; ++i) {
+
+      // Find the joint angles for the next waypoint, starting from the last
+      // waypoint, and save into the position vector.
+      if (end_tip_directions.cols() > 0) {
```

Comment:
Preferably, this conditional branch should be hoisted outside of a loop, considering the if () branch is evaluated above

Reply:
fixed